presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchMetadata.java

```diff
@@ -0,0 +1,169 @@
+    {
+        ElasticsearchTableHandle handle = checkType(table, ElasticsearchTableHandle.class, "table");
+        checkArgument(handle.getConnectorId().equals(connectorId), "tableHandle is not for this connector");
+        SchemaTableName tableName = new SchemaTableName(handle.getSchemaName(), handle.getTableName());
+        return getTableMetadata(session, tableName);
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, String schemaNameOrNull)
+    {
+        return client.listTables(schemaNameOrNull);
+    }
+
+    @Override
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        ElasticsearchTableHandle handle = checkType(tableHandle, ElasticsearchTableHandle.class, "tableHandle");
+        checkArgument(handle.getConnectorId().equals(connectorId), "tableHandle is not for this connector");
+
+        ElasticsearchTableDescription table = client.getTable(handle.getSchemaName(), handle.getTableName());
+        if (table == null) {
+            throw new TableNotFoundException(handle.toSchemaTableName());
+        }
+
+        ImmutableMap.Builder<String, ColumnHandle> columnHandles = ImmutableMap.builder();
+        int index = 0;
+        for (ColumnMetadata column : table.getColumnsMetadata()) {
+            ElasticsearchColumnMetadata metadata = (ElasticsearchColumnMetadata) column;
+            columnHandles.put(metadata.getName(), new ElasticsearchColumnHandle(connectorId, column.getName(), column.getType(), metadata.getJsonPath(), metadata.getJsonType(), index, metadata.isList()));
+            index++;
+        }
+        return columnHandles.build();
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        checkType(tableHandle, ElasticsearchTableHandle.class, "tableHandle");
+        return checkType(columnHandle, ElasticsearchColumnHandle.class, "columnHandle").getColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        requireNonNull(prefix, "prefix is null");
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName tableName : listTables(session, prefix)) {
+            ConnectorTableMetadata tableMetadata = getTableMetadata(session, tableName);
+            // table can disappear during listing operation
+            if (tableMetadata != null) {
+                columns.put(tableName, tableMetadata.getColumns());
+            }
+        }
+        return columns.build();
+    }
+
+    private ConnectorTableMetadata getTableMetadata(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (!listSchemaNames(session).contains(tableName.getSchemaName())) {
```

Comment:
See my comment above about whether this check on L151 is really needed.

Reply:
will check